src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -0,0 +1,136 @@
+open Core_kernel
+open Async_kernel
+
+type 'a t =
+  { mvar: 'a option Mvar.Read_only.t
+  ; mutable cache: 'a option
+  ; mvar_pipe_handle: 'a option Pipe.Reader.t
+  ; mutable reader_id: int
+  ; pipes:
+      ( 'a option Strict_pipe.Reader.t
+      * ( 'a option
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t )
+      Int.Table.t }
+
+let create mvar =
+  let mvar_pipe_handle = Mvar.pipe_when_ready mvar in
+  let t =
+    { mvar
+    ; cache= Mvar.peek mvar |> Option.join
+    ; mvar_pipe_handle
+    ; reader_id= 0
+    ; pipes= Int.Table.create () }
+  in
+  don't_wait_for
+    (Pipe.iter_without_pushback t.mvar_pipe_handle ~f:(fun a ->
+         t.cache <- a ;
+         Int.Table.iter t.pipes ~f:(fun (_, w) -> Strict_pipe.Writer.write w a)
```

Comment:
Looking down below, actually, you aren't using the strict_pipes correctly. You always create a buffered pipe, which is synchronous, but you expose both iter with and without pushbacks, which have no real difference on synchronous pipes since pushback is always ignored. You should write this so that either: A) callers can choose the type of pipe they want (which can be done either at the creation level or even at a per pipe function level if you wanted to get crafty, though it's totally doable), or B) this module only exposes the relevant pipe consumption method so that callers don't make assumptions that pushback is being considered when it isn't.